framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -555,56 +628,131 @@ private AnnotationMirror createNumberAnnotationMirror(List<Number> values) {
                 }
             }
         }
+
+        // Refine the values of the operands in the store based on the result of the comparison.
+
+        List<Number> thenLeftVals = new ArrayList<>();
+        List<Number> elseLeftVals = new ArrayList<>();
+
+        List<Number> thenRightVals = new ArrayList<>();
+        List<Number> elseRightVals = new ArrayList<>();
+
+        // Because resultValues contains the (boolean) results of each comparison (i.e. the product of the number
+        // of elements in each list), we have to iterate through parts of the list at a time in order to find all
+        // the results of comparisons involving a particular value in the left list.
+        for (int i = 0; i < lefts.size(); i++) {
+            for (int j = 0; j < rights.size(); j++) {
+                if (resultValues.get(j + i * rights.size())) {
```

Comment:
While I agree that the comment on this loop could be better (I'll rewrite it), I don't think it makes sense to try to do this above. That would require duplicating the same three or four lines of code for each type of comparison (something like: if the result of the comparison was true, put it in this list, else put it in that list). I think the single pair of for loops is clearer, even if the indexing is difficult to grasp. But I think we can make it easier to understand with a better comment.